scripts/db-heal-test.py

```diff
@@ -0,0 +1,285 @@
+process.wait()
+
+if has_failed:
+    sys.exit(1)
+
+print("Testing db created, starting loop.")
+
+for i in range(1, number_of_iterations + 1):
+
+    # Find working dir - Aida copies db-src
+    working_db = find_working_db(working_dir)
+    archive = os.path.join(working_db, 'archive')
+    live = os.path.join(working_db, 'live')
+
+    # Dump carmen's logs into a file to avoid spamming
+    c = open(carmen_log_file, 'w')
+
+    # Find last checkpoint block
+    latest_checkpoint = get_latest_checkpoint_from_info()
+    print(f"Using checkpoint on block {latest_checkpoint}.")
+
+    # Restore Archive
+    result = subprocess.run(
+        ['go', 'run', './database/mpt/tool', 'reset', '--force-unlock', str(archive), str(latest_checkpoint)],
+        stdout=c,
+        stderr=c,
+        cwd=carmen_path)
+    if has_program_failed(result.returncode, c):
+        # Next error is fatal
+        has_failed = True
+        break
+
+    # Export genesis to restore LiveDB
+    print(f"Restoration complete. Exporting LiveDB genesis.")
+    result = subprocess.run(
+        ['go', 'run', './database/mpt/tool', 'export', '--block', str(latest_checkpoint), str(archive), str(genesis)],
+        stdout=c,
+        stderr=c,
+        cwd=carmen_path)
+    if has_program_failed(result.returncode, c):
+        has_failed = True
+        break
+
+    # Restore LiveDB
+    print("Export complete. Applying LiveDB genesis.")
+    shutil.rmtree(live)
+
+    result = subprocess.run(
+        ['go', 'run', './database/mpt/tool', 'import-live-db', str(genesis), str(live)],
+        stdout=c,
+        stderr=c,
+        cwd=carmen_path)
+    if has_program_failed(result.returncode, c):
+        has_failed = True
+        break
+
+    print(f"Iteration {i}/{number_of_iterations}")
+    # We restored to block X, although we need to start the app at +1 block because X is already done
+    first_block = latest_checkpoint + 1
+
+    print("Restarting Aida...")
+    command = [
+        str(binary_path), 'substate', '--validate',
+        '--db-tmp', working_dir, '--carmen-schema', '5', '--db-impl', 'carmen',
+        '--aida-db', aida_db_path, '--no-heartbeat-logging', '--track-progress',
+        '--archive', '--archive-variant', 's5', '--archive-query-rate', '200',
+        '--carmen-checkpoint-interval', str(checkpoint_granularity), '--db-src',
+        str(working_db), '--skip-priming', '--tracker-granularity',
+        str(checkpoint_granularity), str(first_block), str(last_block)
+    ]
+
+    with open(aida_log_file, 'w') as f:
+        process = subprocess.Popen(command, stdout=f, stderr=subprocess.STDOUT)
+
+    # Start monitoring the log file
+    has_failed = check_aida_log(window, latest_checkpoint)
+
+    # Wait for the command to complete
+    process.wait()
+
+    if has_failed:
+        break
+
+# Clear anything leftover
+print(f"Clearing work directory {working_dir}.")
+shutil.rmtree(working_dir, ignore_errors=True)
```

Comment:
To simplify debugging it would be good to only clear the data if the run was a success. Keeping the working directory in case of failures enables the inspection of the failure state.

If you do not like to keep the data around in any potential failure state, maybe there can be an optional `--keep-data` flag

Reply:
The script now does not delete any data from `working-dir` if it fails.